### PR TITLE
#161716604 Build Out the Endpoint for a User to Create a Parcel Delivery Order

### PR DESCRIPTION
**Pull request by YvesIraguha**
### What does this PR do?
Have an API for creating a delivery order,  lint the codes for better readability, and transpile codes to es5. 

### Description of the task to be completed?
* Have the endpoint for creating order working properly
* Transpile codes to es5
* Lint the codes for better readability
* Separate the business logic from routes definition
* Create multiple test files with different test cases
* Create the data model for a parcel delivery order

### How should this be manually tested?
_After cloning the repo, change directory to API and run_:
* `npm run build` to transpile to es6 
* `npm run test` ` to run tests 
* `npm start` to start the local server 

### Background context
This is the application of the feedback got from my reviewer 

### Relevant pivotal tracker stories 

#161716604
#161786081
#161717052
#161716963
#161750279
#161776656
#161716856
#161773793
#161716750

### CODE REVIEW

File: API/src/models/parcel.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Class for defining a parcel model
+class Parcel {
+  constructor(id, name, origin, destination, weight, userId) {
+    this.id = id;
+    this.name = name;
+    this.origin = origin;
+    this.destination = destination;
+    this.weight = weight;
+    this.userId = userId;
+	    this.setPrice();
```
Comment (johnseremba) [inline API/src/models/parcel.js]: Remove this white space

File: API/src/models/parcel.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  setPrice() {
+    if (typeof this.weight === 'number') {
```
Comment (johnseremba) [inline API/src/models/parcel.js]: Leverage on javascript's truthy and falsy concepts.
Could be refactored to if (Number(this.weight)) ...

File: API/src/routes/parcels.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const parcelsRouter = express.Router();
+
+// redirect to the apis when on heroku.
+parcelsRouter.get('/', (req, res) => {
+  res.redirect('/parcels');
+});
+
+// fetch a parcel by id
+parcelsRouter.get('/parcels/:id', controllers.fetchParcelById);
+// Route for accepting data from the parcel creation.
```
Comment (johnseremba) [inline API/src/routes/parcels.js]: Consider adding spaces every after a router definition.

File: API/src/controlers/parcelControllers.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+// create parcel
+const createParcel = (req, res) => {
+  const {
+    name, origin, destination, weight, userId,
+  } = req.body;
+  const existingOrder = orders.find(order => order.name === name);
+  if (existingOrder === undefined) {
+    const id = orders.length + 1;
+    const fieldsValidation = /[a-zA-Z]+/;
+    if (!origin) {
```
Comment (johnseremba) [inline API/src/controlers/parcelControllers.js]: This block of code is all handling required fields.
We could instead use a generic response message for the required fields.

```
if (!weight || !name || !destination) { 
  res.send({ message: 'Please provide all the required fields!' });
}
```

File: API/src/controlers/parcelControllers.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  res.send(orders);
+};
+
+// cancel a delivery order
+const cancelDeliveryOrder = (req, res) => {
+  const parcelId = parseInt(req.params.id);
+  const parcel = orders.find(order => order.id === parcelId);
+  if (parcel) {
+    orders.splice(orders.indexOf(parcel), 1);
+    parcel.status = 'Canceled';
```
Comment (johnseremba) [inline API/src/controlers/parcelControllers.js]: Check the spelling of cancelled.

File: API/src/controlers/parcelControllers.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    orders.push(parcel);
+    res.send({ message: 'Order successfully canceled', parcel });
+  } else {
+    res.send({ message: 'Invalid Id' });
```
Comment (johnseremba) [inline API/src/controlers/parcelControllers.js]: change message to : "Invalid id"

**Review by johnseremba — CHANGES_REQUESTED** [review]

**Comment by YvesIraguha** [comment]
The routes has been updated and the database was configured in ch-setup-database branch

**Pull request closed without merging**